pkg/controller/certificates/clustertrustbundlepublisher/publisher.go

```diff
@@ -0,0 +1,226 @@
+	defer logger.Info("Shutting down ClusterTrustBundle CA cert publisher controller")
+
+	go p.ctbInformer.Run(ctx.Done())
+
+	if !cache.WaitForNamedCacheSync("cluster trust bundle", ctx.Done(), p.ctbListerSynced) {
+		return
+	}
+
+	// init the signer syncer
+	p.queue.Add("")
+	go wait.UntilWithContext(ctx, p.runWorker(), time.Second)
+
+	<-ctx.Done()
+}
+
+func (p *ClusterTrustBundlePublisher) runWorker() func(context.Context) {
+	return func(ctx context.Context) {
+		for p.processNextWorkItem(ctx) {
+		}
+	}
+}
+
+// processNextWorkItem deals with one key off the queue. It returns false when
+// it's time to quit.
+func (p *ClusterTrustBundlePublisher) processNextWorkItem(ctx context.Context) bool {
+	key, quit := p.queue.Get()
+	if quit {
+		return false
+	}
+	defer p.queue.Done(key)
+
+	if err := p.syncClusterTrustBundle(ctx); err != nil {
+		utilruntime.HandleError(fmt.Errorf("syncing %q failed: %w", key, err))
+		p.queue.AddRateLimited(key)
+		return true
+	}
+
+	p.queue.Forget(key)
+	return true
+}
+
+func (p *ClusterTrustBundlePublisher) syncClusterTrustBundle(ctx context.Context) (err error) {
+	startTime := time.Now()
+	defer func() {
+		recordMetrics(startTime, err)
+		klog.FromContext(ctx).V(4).Info("Finished syncing ClusterTrustBundle", "signerName", p.signerName, "elapsedTime", time.Since(startTime))
+	}()
+
+	caBundle := string(p.ca.CurrentCABundleContent())
+	bundleName := constructBundleName(p.signerName, []byte(caBundle))
+
+	bundle, err := p.ctbLister.Get(bundleName)
+	if apierrors.IsNotFound(err) {
+		_, err = p.client.CertificatesV1alpha1().ClusterTrustBundles().Create(ctx, &certificatesv1alpha1.ClusterTrustBundle{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: bundleName,
+			},
+			Spec: certificatesv1alpha1.ClusterTrustBundleSpec{
+				SignerName:  p.signerName,
+				TrustBundle: caBundle,
+			},
+		}, metav1.CreateOptions{})
+	} else if err == nil && bundle.Spec.TrustBundle != caBundle {
+		bundle = bundle.DeepCopy()
+		bundle.Spec.TrustBundle = caBundle
+		_, err = p.client.CertificatesV1alpha1().ClusterTrustBundles().Update(ctx, bundle, metav1.UpdateOptions{})
+	}
+
+	if err != nil {
+		return err
+	}
+
+	signerTrustBundles, err := p.ctbLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, bundleObject := range signerTrustBundles {
+		if bundleObject.Name == bundleName {
+			continue
+		}
+
+		if err := p.client.CertificatesV1alpha1().ClusterTrustBundles().Delete(ctx, bundleObject.Name, metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func setupSignerNameFilteredCTBInformer(client clientset.Interface, signerName string) cache.SharedIndexInformer {
+	return certinformers.NewFilteredClusterTrustBundleInformer(client, 0, cache.Indexers{},
+		func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("spec.signerName", signerName).String()
+		})
+}
+
+func constructBundleName(signerName string, bundleBytes []byte) string {
```

Comment:
Can you elaborate on why you construct the bundle name using a hash of the content, rather than just a fixed derivation from the signer name?  It seems like it reads to a bunch of potentially-harmful races where multiple CTBs for the same signer are live at the same time.

Reply:
> Can you elaborate on why you construct the bundle name using a hash of the content, rather than just a fixed derivation from the signer name

This is per the KEP, so that we don't make the name an API.